lib/Translation/EmitHLSCpp.cpp

```diff
@@ -692,6 +698,69 @@ void ModuleEmitter::emitConstBuffer(ConstBufferOp op) {
   emitArrayDirectives(op.getResult());
 }
 
+/// Library Ip emitter
+void ModuleEmitter::emitLibraryIp(InstanceOp op) {
+  indent();
+  /// Get Lib name. (This is not used, but left for future updates)
+  // auto libName = op->getAttrOfType<mlir::SymbolRefAttr>("name");
+  // os << libName.getRootReference().getValue().str();
+
+  // Get Ip name, print Ip name.
+  auto ipName = op->getAttrOfType<mlir::SymbolRefAttr>("name");
+  os << ipName.getNestedReferences()[0].getValue().str();
+
+  // Emit template.
+  os << "<";
+  auto allTemplate = op.getTemplates();
+  for (unsigned i = 0; i < allTemplate.size(); ++i) {
+    if (auto curAttr = allTemplate[i].dyn_cast<TypeAttr>()) {
+
+      // If template is a TPYE, print out the corresponding type string
+      if (auto floatType = curAttr.getValue().dyn_cast<FloatType>()) {
+        os << "float";
+      } else if (auto indexType = curAttr.getValue().dyn_cast<IndexType>()) {
+        os << "Int";
+      }
+
+      // If template is a number, print the number. (For now, support integer
+      // only)
+    } else if (auto curAttr = allTemplate[i].dyn_cast<IntegerAttr>()) {
+      os << curAttr.getValue();
+    }
```

Comment:
It'd be good to have something like `else llvm_unreachable("invalid template parameter");` here.